src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2454,22 +2454,24 @@ void ShenandoahHeap::set_evacuation_reserve_quantities(bool is_valid) {
 }
 
 void ShenandoahHeap::set_concurrent_young_mark_in_progress(bool in_progress) {
-  if (has_forwarded_objects()) {
-    set_gc_state_mask(YOUNG_MARKING | UPDATEREFS, in_progress);
-  } else {
+  assert(!has_forwarded_objects(), "Young marking is not concurrent with evacuation");
+  if (!in_progress && is_concurrent_old_mark_in_progress()) {
+    assert(mode()->is_generational(), "Only generational GC has old marking");
+    // If old-marking is in progress when we turn off YOUNG_MARKING, leave MARKING (and OLD_MARKING) on
     set_gc_state_mask(YOUNG_MARKING, in_progress);
+  } else {
+    set_gc_state_mask(MARKING | YOUNG_MARKING, in_progress);
   }
-
   manage_satb_barrier(in_progress);
 }
 
 void ShenandoahHeap::set_concurrent_old_mark_in_progress(bool in_progress) {
-  if (has_forwarded_objects()) {
-    set_gc_state_mask(OLD_MARKING | UPDATEREFS, in_progress);
-  } else {
+  if (!in_progress && is_concurrent_young_mark_in_progress()) {
+    // If young-marking is in progress when we turn off OLD_MARKING, leave MARKING (and YOUNG_MARKING) on
```

Comment:
Similarly, assert that `MARKING` is set here?

Reply:
I'm making this refinement also.